core/src/main/java/jenkins/model/PeepholePermalink.java

```diff
@@ -94,115 +90,216 @@ protected File getPermalinkFile(Job<?, ?> job) {
-            n = cache.getOrDefault(getId(), 0);
+        return ExtensionList.lookupFirst(Cache.class).get(job, getId()).resolve(this, job, getId());
+    }
+
+    /**
+     * Start from the build 'b' and locate the build that matches the criteria going back in time
+     */
+    @CheckForNull
+    private Run<?, ?> find(@CheckForNull Run<?, ?> b) {
+        while (b != null && !apply(b)) {
+            b = b.getPreviousBuild();
         }
-        if (n == RESOLVES_TO_NONE) {
-            return null;
+        return b;
+    }
+
+    /**
+     * Remembers the value 'n' in the cache for future {@link #resolve(Job)}.
+     */
+    protected void updateCache(@NonNull Job<?, ?> job, @CheckForNull Run<?, ?> b) {
+        ExtensionList.lookupFirst(Cache.class).put(job, getId(), b != null ? new Cache.Some(b.getNumber()) : Cache.NONE);
+    }
+
+    /**
+     * Persistable cache of peephole permalink targets.
+     */
+    @Restricted(Beta.class)
+    public interface Cache extends ExtensionPoint {
+
+        /** Cacheable target of a permalink. */
+        sealed interface PermalinkTarget extends Serializable {
+
+            /**
+             * Implementation of {@link #resolve(Job)}.
+             * This may update the cache if it was missing or found to be invalid.
+             */
+            @Restricted(NoExternalUse.class)
+            @CheckForNull
+            Run<?, ?> resolve(@NonNull PeepholePermalink pp, @NonNull Job<?, ?> job, @NonNull String id);
+
+            /**
+             * Partial implementation of {@link #resolve(PeepholePermalink, Job, String)} when searching.
+             * @param b if set, the newest build to even consider when searching
+             */
+            @Restricted(NoExternalUse.class)
+            @CheckForNull
+            default Run<?, ?> search(@NonNull PeepholePermalink pp, @NonNull Job<?, ?> job, @NonNull String id, @CheckForNull Run<?, ?> b) {
+                if (b == null) {
+                    // no cache
+                    b = job.getLastBuild();
+                }
+                // start from the build 'b' and locate the build that matches the criteria going back in time
+                b = pp.find(b);
+                pp.updateCache(job, b);
+                return b;
+            }
+
         }
-        Run<?, ?> b;
-        if (n > 0) {
-            b = job.getBuildByNumber(n);
-            if (b != null && apply(b)) {
-                return b; // found it (in the most efficient way possible)
+
+        /**
+         * The cache entry for this target is missing.
+         */
+        record Unknown() implements PermalinkTarget {
+            @Override
+            public Run<?, ?> resolve(PeepholePermalink pp, Job<?, ?> job, String id) {
+                return search(pp, job, id, null);
             }
-        } else {
-            b = null;
         }
 
-        // the cache is stale. start the search
-        if (b == null) {
-            b = job.getNearestOldBuild(n);
```

Comment:
It took me a while to realize that this was sometimes called with `n == 0` in which case the return value of `getNearestOldBuild` was unconditionally `null`, so the clause had no effect (`b == null` before and after). Now this method is called only when processing `Some`, with a positive argument.